File: src/lib/api/signforms.ts
/**
 * SignForms API Service
 * =====================
 * Connects the frontend to the /signforms backend routes.
 */

import { API_URL, authHeaders, authFetch } from './client';

/* ─── Types ─── */

export interface SignFormData {
    id: string;
    owner_id: string;
    template_id: string;
    name: string;
    description: string | null;
    status: 'active' | 'inactive' | 'completed';
    responses_count: number;
    max_responses: number | null;
    url: string;
    expiry_date: string | null;
    created_at: string;
    updated_at: string;
}

export interface SignFormResponseEntry {
    id: string;
    sign_form_id: string;
    signer_name: string;
    signer_email: string;
    status: 'completed' | 'pending' | 'expired';
    recipient_id: string | null;
    field_values: Record<string, unknown> | null;
    signed_at: string | null;
    created_at: string;
}

export interface SignFormDetailData extends SignFormData {
    responses_list: SignFormResponseEntry[];
}

export interface SignFormListResponse {
    sign_forms: SignFormData[];
    total: number;
}

/* ─── API Calls ─── */

export async function createSignForm(data: {
    template_id: string;
    name: string;
    description?: string;
    max_responses?: number;
    expiry_date?: string;
}): Promise<SignFormData> {
    const res = await authFetch(`${API_URL}/signforms`, {
        method: 'POST',
        headers: authHeaders(),
        body: JSON.stringify(data),
    });
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Failed to create sign form');
    return json;
}

export async function listSignForms(params?: {
    status?: string;
    search?: string;
    limit?: number;
    offset?: number;
}): Promise<SignFormListResponse> {
    const query = new URLSearchParams();
    if (params?.status) query.set('status_filter', params.status);
    if (params?.search) query.set('search', params.search);
    if (params?.limit) query.set('limit', String(params.limit));
    if (params?.offset) query.set('offset', String(params.offset));

    const url = `${API_URL}/signforms${query.toString() ? `?${query}` : ''}`;
    const res = await authFetch(url, { headers: authHeaders() });
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Failed to list sign forms');
    return json;
}

export async function getSignForm(id: string): Promise<SignFormDetailData> {
    const res = await authFetch(`${API_URL}/signforms/${id}`, {
        headers: authHeaders(),
    });
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Failed to get sign form');
    return json;
}

export async function updateSignForm(
    id: string,
    data: {
        name?: string;
        description?: string;
        status?: 'active' | 'inactive' | 'completed';
        max_responses?: number;
        expiry_date?: string;
    },
): Promise<SignFormData> {
    const res = await authFetch(`${API_URL}/signforms/${id}`, {
        method: 'PATCH',
        headers: authHeaders(),
        body: JSON.stringify(data),
    });
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Failed to update sign form');
    return json;
}

export async function deleteSignForm(id: string): Promise<void> {
    const res = await authFetch(`${API_URL}/signforms/${id}`, {
        method: 'DELETE',
        headers: authHeaders(),
    });
    if (!res.ok) {
        const json = await res.json();
        throw new Error(json.detail || 'Failed to delete sign form');
    }
}

export async function getPublicSignForm(urlSlug: string): Promise<SignFormData> {
    const res = await fetch(`${API_URL}/signforms/public/${urlSlug}`);
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Sign form not found or inactive');
    return json;
}

/* ─── Public Template Content (for signing interface) ─── */

export interface PublicTemplateData {
    template_name: string;
    content: string;
    fields_config: {
        sendInOrder: boolean;
        recipients: {
            id: string;
            name: string;
            email: string;
            role: string;
            color: string;
            action: 'sign' | 'view' | 'approve';
            notifyVia: string;
        }[];
        fields: {
            id: string;
            type: string;
            label: string;
            recipientId: string;
            x: number;
            y: number;
            width: number;
            height: number;
            required: boolean;
            page: number;
            placeholder?: string;
        }[];
    } | null;
}

export async function getPublicTemplate(urlSlug: string): Promise<PublicTemplateData> {
    const res = await fetch(`${API_URL}/signforms/public/${urlSlug}/template`);
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Could not load template');
    return json;
}

/* ─── Submit Signing Response (public, no auth) ─── */

export interface SubmitResponsePayload {
    signer_name: string;
    signer_email: string;
    field_values?: Record<string, unknown>;
    recipient_id?: string;
}

export interface SubmitResponseResult {
    id: string;
    sign_form_id: string;
    signer_name: string;
    signer_email: string;
    status: string;
    field_values: Record<string, unknown> | null;
    signed_at: string | null;
    created_at: string;
}

export async function submitSignFormResponse(
    urlSlug: string,
    data: SubmitResponsePayload,
): Promise<SubmitResponseResult> {
    const res = await fetch(`${API_URL}/signforms/public/${urlSlug}/respond`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data),
    });
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Failed to submit response');
    return json;
}

/* ─── Pending Sign Forms for Current User ─── */

export interface PendingSignFormItem {
    sign_form_id: string;
    sign_form_name: string;
    sign_form_url: string;
    template_name: string;
    template_content: string | null;
    owner_name: string;
    recipient_role: string;
    created_at: string;
}

export interface PendingSignFormsResponse {
    pending: PendingSignFormItem[];
    total: number;
}

export async function getPendingForMe(): Promise<PendingSignFormsResponse> {
    const res = await authFetch(`${API_URL}/signforms/pending-for-me`, {
        headers: authHeaders(),
    });
    const json = await res.json();
    if (!res.ok) throw new Error(json.detail || 'Failed to fetch pending sign forms');
    return json;
}

File: backend/crud/signforms.py
"""
SignForms CRUD Operations
=========================
Database operations for sign forms.
"""

import uuid
from typing import Optional, List
from supabase import Client


def _generate_url_slug() -> str:
    """Generate a short unique URL slug for the sign form."""
    return uuid.uuid4().hex[:12]


def create_signform(
    supabase: Client,
    owner_id: str,
    template_id: str,
    name: str,
    description: str = "",
    max_responses: Optional[int] = None,
    expiry_date: Optional[str] = None,
) -> dict:
    """
    Create a new sign form linked to a template.
    Generates a unique URL slug automatically.
    """
    row = {
        "owner_id": owner_id,
        "template_id": template_id,
        "name": name,
        "description": description or "",
        "status": "active",
        "responses_count": 0,
        "url": _generate_url_slug(),
    }

    if max_responses is not None:
        row["max_responses"] = max_responses

    if expiry_date is not None:
        row["expiry_date"] = expiry_date

    response = (
        supabase.table("sign_forms")
        .insert(row)
        .execute()
    )
    return response.data[0] if response.data else None


def get_signform_by_id(supabase: Client, signform_id: str) -> Optional[dict]:
    """
    Get a single sign form by its ID.
    """
    response = (
        supabase.table("sign_forms")
        .select("*")
        .eq("id", signform_id)
        .single()
        .execute()
    )
    return response.data


def get_signform_by_url(supabase: Client, url_slug: str) -> Optional[dict]:
    """
    Get a sign form by its public URL slug (for public access).
    """
    response = (
        supabase.table("sign_forms")
        .select("*")
        .eq("url", url_slug)
        .single()
        .execute()
    )
    return response.data


def get_signforms_by_owner(
    supabase: Client,
    owner_id: str,
    status: Optional[str] = None,
    search: Optional[str] = None,
    limit: int = 50,
    offset: int = 0,
) -> List[dict]:
    """
    List all sign forms belonging to a user.
    Supports optional filtering by status and search by name.
    """
    query = (
        supabase.table("sign_forms")
        .select("*")
        .eq("owner_id", owner_id)
    )

    if status:
        query = query.eq("status", status)

    if search:
        query = query.ilike("name", f"%{search}%")

    query = query.order("updated_at", desc=True).range(offset, offset + limit - 1)

    response = query.execute()
    return response.data or []


def count_signforms_by_owner(supabase: Client, owner_id: str) -> int:
    """
    Return the total number of sign forms owned by a user.
    """
    response = (
        supabase.table("sign_forms")
        .select("id", count="exact")
        .eq("owner_id", owner_id)
        .execute()
    )
    return response.count or 0


def update_signform(supabase: Client, signform_id: str, updates: dict) -> Optional[dict]:
    """
    Update a sign form. Only non-None fields in `updates` are applied.
    """
    clean = {k: v for k, v in updates.items() if v is not None}
    if not clean:
        return get_signform_by_id(supabase, signform_id)

    response = (
        supabase.table("sign_forms")
        .update(clean)
        .eq("id", signform_id)
        .execute()
    )
    return response.data[0] if response.data else None


def delete_signform(supabase: Client, signform_id: str) -> bool:
    """
    Delete a sign form by ID. Returns True if a row was deleted.
    """
    response = (
        supabase.table("sign_forms")
        .delete()
        .eq("id", signform_id)
        .execute()
    )
    return bool(response.data)


# ─── Sign Form Responses ─────────────────────────────────────────────────────

def get_responses_by_signform(
    supabase: Client,
    signform_id: str,
    status: Optional[str] = None,
    limit: int = 100,
    offset: int = 0,
) -> List[dict]:
    """
    List all responses for a given sign form.
    """
    query = (
        supabase.table("sign_form_responses")
        .select("*")
        .eq("sign_form_id", signform_id)
    )

    if status:
        query = query.eq("status", status)

    query = query.order("created_at", desc=True).range(offset, offset + limit - 1)

    response = query.execute()
    return response.data or []


def count_responses_by_signform(supabase: Client, signform_id: str) -> int:
    """
    Return the total number of responses for a sign form.
    """
    response = (
        supabase.table("sign_form_responses")
        .select("id", count="exact")
        .eq("sign_form_id", signform_id)
        .execute()
    )
    return response.count or 0


def create_response(
    supabase: Client,
    sign_form_id: str,
    signer_name: str,
    signer_email: str,
    recipient_id: str = "",
    field_values: Optional[dict] = None,
    ip_address: str = "",
) -> dict:
    """
    Create a response entry for a sign form (public signing endpoint).
    """
    row = {
        "sign_form_id": sign_form_id,
        "signer_name": signer_name,
        "signer_email": signer_email,
        "status": "pending",
        "recipient_id": recipient_id,
        "field_values": field_values or {},
        "ip_address": ip_address,
    }

    response = (
        supabase.table("sign_form_responses")
        .insert(row)
        .execute()
    )
    return response.data[0] if response.data else None


def update_response(supabase: Client, response_id: str, updates: dict) -> Optional[dict]:
    """
    Update a sign form response (e.g., mark as completed with field values).
    """
    clean = {k: v for k, v in updates.items() if v is not None}
    if not clean:
        return None

    response = (
        supabase.table("sign_form_responses")
        .update(clean)
        .eq("id", response_id)
        .execute()
    )
    return response.data[0] if response.data else None
